File: python/graph_extra.py
from typing import List, Tuple


class DSU:
    def __init__(self, n: int):
        self.parent = list(range(n))

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, a: int, b: int) -> bool:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        self.parent[ra] = rb
        return True


def prim(n: int, graph: List[List[Tuple[int, int]]], start: int = 0) -> Tuple[int, int]:
    inf = 10**18
    dist = [inf] * n
    used = [False] * n
    dist[start] = 0

    weight = 0
    count = 0
    for _ in range(n):
        u = -1
        for v in range(n):
            if not used[v] and (u == -1 or dist[v] < dist[u]):
                u = v
        if u == -1 or dist[u] == inf:
            break
        used[u] = True
        weight += dist[u]
        count += 1

        for to, w in graph[u]:
            if not used[to] and w < dist[to]:
                dist[to] = w

    return weight, count


def spfa(n: int, graph: List[List[Tuple[int, int]]], src: int) -> List[int]:
    inf = 10**18
    dist = [inf] * n
    in_q = [False] * n
    q = []

    dist[src] = 0
    q.append(src)
    in_q[src] = True

    i = 0
    while i < len(q):
        u = q[i]
        i += 1
        in_q[u] = False

        for to, w in graph[u]:
            if dist[u] + w < dist[to]:
                dist[to] = dist[u] + w
                if not in_q[to]:
                    q.append(to)
                    in_q[to] = True

    return dist

File: ts/src/graph_extra.ts
type Edge = { to: number; w: number };

type WEdge = { u: number; v: number; w: number };

export class DSU {
  private parent: number[];

  constructor(n: number) {
    this.parent = Array.from({ length: n }, (_, i) => i);
  }

  find(x: number): number {
    if (this.parent[x] !== x) this.parent[x] = this.find(this.parent[x]);
    return this.parent[x];
  }

  union(a: number, b: number): boolean {
    const ra = this.find(a);
    const rb = this.find(b);
    if (ra === rb) return false;
    this.parent[ra] = rb;
    return true;
  }
}

export function prim(n: number, g: Edge[][], start = 0): { weight: number; count: number } {
  const inf = Number.POSITIVE_INFINITY;
  const dist = new Array<number>(n).fill(inf);
  const used = new Array<boolean>(n).fill(false);
  dist[start] = 0;

  let weight = 0;
  let count = 0;

  for (let i = 0; i < n; i++) {
    let u = -1;
    for (let v = 0; v < n; v++) {
      if (!used[v] && (u === -1 || dist[v] < dist[u])) u = v;
    }
    if (u === -1 || dist[u] === inf) break;
    used[u] = true;
    weight += dist[u];
    count++;

    for (const e of g[u]) {
      if (!used[e.to] && e.w < dist[e.to]) dist[e.to] = e.w;
    }
  }

  return { weight, count };
}

export function spfa(n: number, g: Edge[][], src: number): number[] {
  const inf = Number.POSITIVE_INFINITY;
  const dist = new Array<number>(n).fill(inf);
  const inQ = new Array<boolean>(n).fill(false);
  const q: number[] = [];

  dist[src] = 0;
  q.push(src);
  inQ[src] = true;

  for (let i = 0; i < q.length; i++) {
    const u = q[i];
    inQ[u] = false;

    for (const e of g[u]) {
      if (dist[u] + e.w < dist[e.to]) {
        dist[e.to] = dist[u] + e.w;
        if (!inQ[e.to]) {
          q.push(e.to);
          inQ[e.to] = true;
        }
      }
    }
  }

  return dist;
}

export function kruskalWithDsu(n: number, edges: WEdge[]): { weight: number; count: number } {
  const dsu = new DSU(n);
  let weight = 0;
  let count = 0;

  edges.sort((a, b) => a.w - b.w);
  for (const e of edges) {
    if (dsu.union(e.u, e.v)) {
      weight += e.w;
      count++;
      if (count === n - 1) break;
    }
  }

  return { weight, count };
}
